std/utf.d

```diff
@@ -3092,77 +3093,74 @@ enum dchar replacementDchar = '\uFFFD';
  * one while iterating over the resulting range will give nonsensical results.
  *
  * Params:
- *      r = an input range of characters, or an array of characters
+ *      r = an input range of characters (including strings) or a type that
+ *          implicitly converts to a string type.
  * Returns:
- *     If `r` is not an auto-decodable string, then `r` is returned.
+ *     If `r` is not an auto-decodable string (i.e. a narrow string or a
+ *     user-defined type that implicits converts to a string type), then `r`
+ *     is returned.
  *
- *      Otherwise, an input range with a length if $(REF isAggregateType, std,traits)
- *      is `true` for `R`. Otherwise, this returns a finite random access range
- *      with slicing.
+ *      Otherwise, `r` is converted to its corresponding string type (if it's
+ *      not already a string) and wrapped in a random-access range where the
+ *      element encoding type of the string (its code unit) is the element type
+ *      of the range, and that range returned. The range has slicing.
+ *
+ *      If `r` is quirky enough to be a struct or class which is an input range
+ *      of characters on its own (i.e. it has the input range API as member
+ *      functions), $(I and) it's implicitly convertible to a string type, then
+ *      `r` is returned, and no implicit conversion takes place.
  * See_Also:
- *      Refer to the $(MREF std, uni) docs for a reference on Unicode terminology.
+ *      Refer to the $(MREF std, uni) docs for a reference on Unicode
+ *      terminology.
  *
  *      For a range that iterates by grapheme cluster (written character) see
  *      $(REF byGrapheme, std,uni).
  */
-auto byCodeUnit(R)(R r) if (isAutodecodableString!R)
-{
-    static struct ByCodeUnitImpl
-    {
-    pure nothrow @nogc:
-
-        @property bool empty() const         { return r.length == 0; }
-        @property auto ref front() inout     { return r[0]; }
-        void popFront()                      { r = r[1 .. $]; }
-        auto ref opIndex(size_t index) inout { return r[index]; }
-
-        @property auto ref back() inout
+auto byCodeUnit(R)(R r)
+    if (isAutodecodableString!R ||
+        isInputRange!R && isSomeChar!(ElementEncodingType!R) ||
+        (is(R == enum) && is(R : const dchar[])))
```

Comment:
It is and it isn't. If anything, it's a great example of why we need to be _very_ careful in defining what on earth "string-like" means.